### PR TITLE
add authorization

### PR DESCRIPTION
**Pull request by iSharovatov1**
# [NewsSiteBack](https://github.com/iSharovatov1/NewsSiteBack)
> Implementation of authorization (server)
### Code Details
* added authorization
### Feature tag selection
Has a `Implementation of authorization (server )` Has it been added [to the tracker](https://trello.com/b/EICN697S/%D1%88%D0%B0%D1%80%D0%BE%D0%B2%D0%B0%D1%82%D0%BE%D0%B2%D0%B8%D0%B3%D0%BE%D1%80%D1%8C%D0%BD%D0%BE%D0%B2%D0%BE%D1%81%D1%82%D0%BD%D0%BE%D0%B9%D1%81%D0%B0%D0%B9%D1%82)?

### CODE REVIEW

File: controllers/auth/signin.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+const randtoken = require('rand-token');
+const bcrypt = require('bcrypt-nodejs');
+require('dotenv').config();
+
+const { Users, Tokens } = require('../../db/models');
+
+const key = process.env.secretKey;
+
+async function signin(req, res) {
+  const { email, password } = req.body;
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: const {body:{email, password}} = req;

File: controllers/auth/signin.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  if (!email || !password) {
+    return res.status(400).send({ msg: 'Please pass email and password.' });
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: `Missed data` or `Email and password required`

File: controllers/auth/signin.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+const bcrypt = require('bcrypt-nodejs');
+require('dotenv').config();
+
+const { Users, Tokens } = require('../../db/models');
+
+const key = process.env.secretKey;
+
+async function signin(req, res) {
+  const { email, password } = req.body;
+  if (!email || !password) {
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: check trimmed values with `''`

File: controllers/auth/signin.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    return res.status(400).send({ msg: 'Please pass email and password.' });
+  }
+  try {
+    const user = await Users.findOne({ where: { email } });
+    if (!user) {
+      return res.status(401).send({
+        message: 'Authentication failed. User not found.',
+      });
+    }
+    bcrypt.compare(password, user.password, async (err, resp) => {
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: move this check to user prototype

Comment (tatiana-dunice) [inline controllers/auth/signin.js]: follow async/await syntax

File: controllers/auth/signin.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      if (resp && !err) {
+        const token = jwt.sign(JSON.parse(JSON.stringify(user)), key, { expiresIn: 86400 * 30 });
+        const refresh = randtoken.uid(255);
+        await Tokens.create({
+          userId: user.id,
+          token: refresh,
+        });
+        const tokens = await Tokens.findAll({ where: { userId: user.id } });
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: you can destroy tokens on logout
or by cron job

File: controllers/auth/signin.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+          if (new Date() - item.createdAt > 60 * 60 * 1000) {
+            await Tokens.destroy({ where: { id: item.id } });
+          }
+        });
+        res.cookie('refresh_token', refresh, {
+          secure: false,
+          maxAge: 60 * 60 * 1000,
+          httpOnly: true,
+        });
+        return res.json({ success: true, token });
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: return user as well

File: controllers/auth/signup.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+async function signup(req, res) {
+  const {
+    firstName,
+    lastName,
+    avatar,
+    email,
+    password,
+  } = req.body;
+  if (!email || !password) {
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: the rest of field shouldn't be validated?  - fix it

File: controllers/auth/signup.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    return res.status(400).send({ msg: 'Please pass email and password.' });
+  }
+  try {
+    const oldUser = await Users.findAll({ where: { email } });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: findOne

File: controllers/auth/signup.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    if (oldUser.length) res.status(409).send('user already exists');
+    const hashPassword = bcrypt.hashSync(password, bcrypt.genSaltSync(10), null);
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: move this to beforeCreate model hook

File: controllers/auth/signup.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      email,
+      password: hashPassword,
+    });
+    const token = jwt.sign(JSON.parse(JSON.stringify(user)), key, { expiresIn: 86400 * 30 });
+    const refresh = randtoken.uid(255);
+    await Tokens.create({
+      userId: user.id,
+      token: refresh,
+    });
+    return res.status(201).cookie('token', refresh).send({ token });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: send user as well

**Review by tatiana-dunice — CHANGES_REQUESTED** [review]

File: controllers/auth/logout.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+const { Tokens } = require('../../db/models');
+
+async function logout(req, res) {
+  const token = req.headers.authorization.split(' ')[1];
+  try {
+    await Tokens.destroy({ where: { token } });
```
Comment (tatiana-dunice) [inline controllers/auth/logout.js]: You should remove cookies as well - you mostly use cookies to check if user aauthorized

File: controllers/auth/logout.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    return res.json({ success: true, token });
```
Comment (tatiana-dunice) [inline controllers/auth/logout.js]: missed status

File: controllers/auth/signin.js [hunk 13]
```diff
@@ -1,16 +1,16 @@
 const jwt = require('jsonwebtoken');
 const randtoken = require('rand-token');
-const bcrypt = require('bcrypt-nodejs');
+const cron = require('node-cron');
 require('dotenv').config();
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: move `require('dotenv').config();` to app.js

File: controllers/auth/signup.js [hunk 14]
```diff
@@ -9,19 +9,21 @@ const key = process.env.secretKey;
...
-    return res.status(400).send({ msg: 'Please pass email and password.' });
+    body: {
+      firstName,
+      lastName,
+      avatar,
+      email,
+      password,
+    },
+  } = req;
+  if (email === '' || password === '' || firstName === '' || lastName === '') {
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: You need to trim values

File: controllers/auth/signup.js [hunk 15]
```diff
@@ -35,7 +37,7 @@ async function signup(req, res) {
       userId: user.id,
       token: refresh,
     });
-    return res.status(201).cookie('token', refresh).send({ token });
+    return res.status(201).cookie('token', refresh).send({ token, user });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: Why don't you set cookies when user signup?

**Review by tatiana-dunice — CHANGES_REQUESTED** [review]

File: controllers/auth/logout.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
+const { Tokens } = require('../../db/models');
+
+async function logout(req, res) {
+  const { cookies: { refreshToken } } = req;
+  try {
+    res.clearCookie('refreshToken');
+    await Tokens.destroy({ where: { token: refreshToken } });
+    return res.status(200).json({ success: true, refreshToken });
```
Comment (tatiana-dunice) [inline controllers/auth/logout.js]: Do you really need to send removed refreshToken to user?

File: controllers/auth/signup.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  const {
+    body: {
+      firstName,
+      lastName,
+      avatar,
+      email,
+      password,
+    },
+  } = req;
+  if (email.trim() === '' || password.trim() === '' || firstName.trim() === '' || lastName.trim() === '') {
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: split into several lines - it's hard to follow this condition

File: controllers/auth/signup.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    return res.status(400).send({ msg: 'Missed data.' });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js:27]: You need to check that email is really email - not a random string - you can use regexp for that

File: controllers/auth/signup.js [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  }
+  const oldUser = await Users.findOne({ where: { email: email.trim() } });
+  if (oldUser) res.status(409).send('user already exists');
+  const hashPassword = bcrypt.hashSync(password, bcrypt.genSaltSync(10), null);
+  try {
+    const user = await Users.create({
+      firstName,
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: you trimed values, but put original strings to create function - fix it

File: controllers/auth/signup.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+      avatar,
+      email,
+      password,
+    },
+  } = req;
+  if (email.trim() === '' || password.trim() === '' || firstName.trim() === '' || lastName.trim() === '') {
+    return res.status(400).send({ msg: 'Missed data.' });
+  }
+  const oldUser = await Users.findOne({ where: { email: email.trim() } });
+  if (oldUser) res.status(409).send('user already exists');
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: wrap code for this condition in `{}` brackets and `return res....`

File: controllers/auth/signup.js [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+  const hashPassword = bcrypt.hashSync(password, bcrypt.genSaltSync(10), null);
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: use beforeCreate hook for user model for that

File: controllers/auth/signup.js [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+  try {
+    const user = await Users.create({
+      firstName,
+      lastName,
+      avatar,
+      email,
+      password: hashPassword,
+    });
+    const token = jwt.sign(JSON.parse(JSON.stringify(user)), key, { expiresIn: 86400 * 30 });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]:
```
const token = jwt.sign(JSON.parse(JSON.stringify(user)), key, { expiresIn: 86400 * 30 });
    const refresh = randtoken.uid(255);
    await Tokens.create({
      userId: user.id,
      token: refresh,
    });
    res.cookie('refreshToken', refresh, {
      secure: false,
      maxAge: 60 * 60 * 1000,
      httpOnly: true,
    });
    return res.status(201).cookie('token', refresh).send({ token, user });
    ```
    this is a code duplication - move to new function

File: controllers/auth/signup.js [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+    await Tokens.create({
+      userId: user.id,
+      token: refresh,
+    });
+    res.cookie('refreshToken', refresh, {
+      secure: false,
+      maxAge: 60 * 60 * 1000,
+      httpOnly: true,
+    });
+    return res.status(201).cookie('token', refresh).send({ token, user });
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: `.cookie` is redundant - you've already set it on the 39th line

**Review by tatiana-dunice — CHANGES_REQUESTED** [review]

File: db/models/users.js [hunk 24]
```diff
@@ -23,5 +24,15 @@ module.exports = (sequelize, DataTypes) => {
...
     modelName: 'Users',
   });
+  Users.prototype.comparePassword = async function (password) {
+    const isCompare = await bcrypt.compare(password, this.password);
+    return isCompare;
+  };
+  Users.addHook(
+    'beforeCreate',
+    (user) => {
+      user.password = bcrypt.hashSync(user.password, bcrypt.genSaltSync(10), null);
```
Comment (tatiana-dunice) [inline db/models/users.js]: what would be happen if user didn't pass password?

File: controllers/auth/signup.js (hunk 22 above)
Comment (tatiana-dunice) [reply controllers/auth/signup.js]: this is a code duplication - move to new function

File: controllers/auth/signup.js [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+    body: {
+      firstName,
+      lastName,
+      avatar,
+      email,
+      password,
+    },
+  } = req;
+  const regexp = /^([A-Za-z0-9_\-\.])+\@([A-Za-z0-9_\-\.])+\.([A-Za-z]{2,4})$/;
+  if (email.trim() === ''
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: reformat the code

File: controllers/auth/signin.js [hunk 26]
```diff
@@ -0,0 +1,42 @@
+const jwt = require('jsonwebtoken');
+const randtoken = require('rand-token');
+
+const { Users, Tokens } = require('../../db/models');
+
+const key = process.env.secretKey;
+
+async function signin(req, res) {
+  const { body: { email, password } } = req;
+  if (email === '' || password === '') {
```
Comment (tatiana-dunice) [inline controllers/auth/signin.js]: trim email

File: controllers/auth/signup.js [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+  const oldUser = await Users.findOne({ where: { email: email.trim() } });
+  if (oldUser) {
+    return res.status(409).send('user already exists');
+  }
+  if (!regexp.test(email)) {
+    return res.status(400).send({ msg: 'Invalid email.' });
+  }
+  try {
+    const user = await Users.create({
+      firstName: firstName.trim(),
```
Comment (tatiana-dunice) [inline controllers/auth/signup.js]: better to create payload with all applicable(trimmed if need) values and use it futher -> in condition and for user creation

File: controllers/news/getAllNews.js [hunk 28]
```diff
@@ -6,6 +6,7 @@ async function getAllNews(req, res) {
       order: [
         ['createdAt', 'DESC'],
       ],
+      include: 'user',
```
Comment (tatiana-dunice) [inline controllers/news/getAllNews.js]: Do you need all user fields in the response? or some exact?

File: db/models/users.js [hunk 29]
```diff
@@ -23,5 +24,15 @@ module.exports = (sequelize, DataTypes) => {
     sequelize,
     modelName: 'Users',
   });
+  Users.prototype.comparePassword = async function (password) {
+    const isCompare = await bcrypt.compare(password, this.password);
```
Comment (tatiana-dunice) [inline db/models/users.js]: try to `return bcrypt.compare(password, this.password);`

File: routes/auth.js [hunk 30]
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+
+const router = express.Router();
+
+const { authController } = require('../controllers');
```
Comment (tatiana-dunice) [inline routes/auth.js:3]: imports should be before variables creation

**Review by tatiana-dunice — CHANGES_REQUESTED** [review]

File: db/models/users.js (hunk 24 above)
Comment (iSharovatov1) [reply db/models/users.js]: user will get error "missed data"

**Review by tatiana-dunice — APPROVED** [review]
